src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -808,6 +874,212 @@ TryDropShardOutsideTransaction(OperationId operationId,
+
+	int querySent = SendRemoteCommand(
+		connection,
+		psprintf("ALTER SUBSCRIPTION %s DISABLE", quote_identifier(subscriptionName)));
+	if (querySent == 0)
+	{
+		ReportConnectionError(connection, ERROR);
+	}
+
+	bool raiseInterrupts = true;
+	PGresult *result = GetRemoteCommandResult(connection, raiseInterrupts);
+	if (!IsResponseOK(result))
+	{
+		char *errorcode = PQresultErrorField(result, PG_DIAG_SQLSTATE);
+		if (errorcode != NULL && strcmp(errorcode, STR_ERRCODE_UNDEFINED_OBJECT) == 0)
+		{
+			/*
+			 * The subscription doesn't exist, so we can return right away.
+			 * This DropSubscription call is effectively a no-op.
+			 */
+			return false;
+		}
+		else
+		{
+			ReportResultError(connection, result, ERROR);
+			PQclear(result);
+			ForgetResults(connection);
+		}
+	}
+
+	PQclear(result);
+	ForgetResults(connection);
+
+	StringInfo alterQuery = makeStringInfo();
+	appendStringInfo(alterQuery,
+					 "ALTER SUBSCRIPTION %s SET (slot_name = NONE)",
+					 quote_identifier(subscriptionName));
+
+	StringInfo dropQuery = makeStringInfo();
+	appendStringInfo(dropQuery,
+					 "DROP SUBSCRIPTION %s",
+					 quote_identifier(subscriptionName));
+
+	List *dropCommandList = list_make2(alterQuery->data, dropQuery->data);
+	bool success = SendOptionalCommandListToWorkerOutsideTransactionWithConnection(
+		connection,
+		dropCommandList);
+
+	return success;
+}
+
+
+/*
+ * TryDropPublicationOutsideTransaction drops the publication with the given name if it
+ * exists.
+ */
+static bool
+TryDropPublicationOutsideTransaction(char *publicationName,
+									 char *nodeName,
+									 int nodePort)
+{
+	int connectionFlags = OUTSIDE_TRANSACTION;
+	MultiConnection *connection = GetNodeUserDatabaseConnection(connectionFlags,
+																nodeName, nodePort,
+																CitusExtensionOwnerName(),
+																NULL);
+	StringInfo dropQuery = makeStringInfo();
+	appendStringInfo(dropQuery,
+					 "DROP PUBLICATION IF EXISTS %s",
+					 quote_identifier(publicationName));
+
+	List *dropCommandList = list_make1(dropQuery->data);
+	bool success = SendOptionalCommandListToWorkerOutsideTransactionWithConnection(
+		connection,
+		dropCommandList);
+
+	return success;
+}
+
+
+/*
+ * TryDropReplicationSlotOutsideTransaction drops the replication slot with the given
+ * name if it exists. It retries if the command fails with an OBJECT_IN_USE error.
+ */
+static bool
+TryDropReplicationSlotOutsideTransaction(char *replicationSlotName,
+										 char *nodeName,
+										 int nodePort)
+{
+	int connectionFlags = OUTSIDE_TRANSACTION;
+	MultiConnection *connection = GetNodeUserDatabaseConnection(connectionFlags,
+																nodeName, nodePort,
+																CitusExtensionOwnerName(),
+																NULL);
+
+	int maxSecondsToTryDropping = 20;
+	bool raiseInterrupts = true;
+	PGresult *result = NULL;
+	bool success = false;
+
+	/* we'll retry in case of an OBJECT_IN_USE error */
+	while (maxSecondsToTryDropping >= 0)
```

Comment:
is this retry logic still necessary? Seems that retries by the background worker should make it fine to drop this code.